src/bartiq/compilation/postprocessing.py

```diff
@@ -42,6 +44,227 @@ def aggregate_resources(
+                # into a new node.
+                for x in chain:
+                    del tree[x]
+
+                chains.append(chain)
+
+                # All offsprings of nodes in chain that don't participate in the merge
+                # have to be re-connected to the new, combined node that we'll create.
+                keys_to_fix |= set(key for key in tree if key in tree[key] in chain)
+
+            # New node is labelled by combining all the participating tuples.
+            # Here we see introducing passthrough (as opposed to only compensating by adding a flow)
+            # is crucial, as otherwise the new_node could have a label that duplicates one of the
+            # existing labels.
+            new_node = sum((node for chain in chains for node in chain), start=())
+            # The new node has highwater equal to sum of highwaters of all chains.
+            highwater_map[new_node] = sum(
+                backend.max(*[highwater_map[node] for node in chain]) for chain in chains if chain
+            )
+
+            # Finally, re-connect the keys that need it.
+            for key in keys_to_fix:
+                tree[key] = new_node
+
+            # New node is a direct descendad of the lca.
+            tree[new_node] = lca
+            # And the new child is a direct descendant of the new node.
+            tree[(child,)] = new_node
+
+    # We added an artificial node to trigger a final summation, after which the tree is
+    # actually linear. However, this node should not be taken account when computing
+    # the final highwater (it's not even in highwater_map) and so we remove it.
+    del tree[(None,)]
+
+    # There are some qubits that are not captured by the above procedure, because they don't
+    # pass through any of children. We adjust for this now. Note that it doesn't matter
+    # which endpoint we take size from, as all those connections are necessary passthrough.
+    passthrough_cost: TExpr[T] = sum(
+        routine.ports[endpoint_1.port_name].size
+        for endpoint_1, endpoint_2 in routine.connections.items()
+        if endpoint_1.routine_name is None and endpoint_2.routine_name is None
+    )
+
+    # Include local ancillae as well.
+    local_ancillae = routine.resources[ancillae_name].value if ancillae_name in routine.resources else 0
+
+    # And finally combine the results.
+    children_highwater = backend.max(*[highwater_map[key] for key in tree])
+    return children_highwater + local_ancillae + passthrough_cost
+
+
+def add_qubit_highwater(
+    routine: CompiledRoutine[T],
+    backend: SymbolicBackend[T],
+    resource_name: str = "qubit_highwater",
+    ancillae_name: str = "local_ancillae",
+) -> CompiledRoutine[T]:
+    """Add information about qubit highwater to the routine.
+
+    Qubit highwater is the number of qubits needed for a particular subroutine, at the place where it's "widest".
```

Comment:
Maybe it's worth mentioning here, that this calculation is an upper bound, as there might be cases, where e.g.: some subroutines need ancilla qubits only for a shot period of time and the same qubits could be reused, but this is not something our model can handle.
?